Make all-reduce input contiguous in `distributed.nn.all_reduce`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144267

Fixes https://github.com/pytorch/pytorch/issues/144060

I confirmed that the unit test fails without the `.contiguous()` fix.

cc @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Commit:
Make all-reduce input contiguous in `distributed.nn.all_reduce`

[ghstack-poisoned]

Skylion007 commented on torch/distributed/nn/functional.py:
```diff
@@ -443,7 +443,7 @@ class _AllReduce(Function):
         ctx.group = group
         ctx.op = op
-        tensor = tensor.clone()
+        tensor = tensor.clone().contiguous()
```
```suggestion
        tensor = tensor.clone(memory_format= torch.contiguous_format)
```
Only need to do one copy

Skylion007 replied:
Actually, this would be a great torchfix check, I see this a few other places in the codebase

Commit:
Update on "Make all-reduce input contiguous in `distributed.nn.all_reduce`"


Fixes https://github.com/pytorch/pytorch/issues/144060

I confirmed that the unit test fails without the `.contiguous()` fix.

cc H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
